### PR TITLE
Remove inferred docstring signature and add basic __signature__ support

### PR DESCRIPTION
**Pull request by maximlt**
Fixes https://github.com/holoviz/param/issues/371 by:

1. removing the automatically inferred *docstring signature* (e.g. `params(x=Number, name=String)`) as IPython would no longer leverage it for parameter autocompletion (as noted in the issue) and because it was wrong when the class was having an `__init__` method with a signature different than `__init__(self, **params)`
2. implementing `Parameterized.__signature__` as a class property that returns a `Signature` built from the class Parameters, **only** when the class (or its superclasses) doesn't implement an `__init__` method that is different than the default one `__init__(self, **params)`.

image

Somewhat surprisingly 2) seemed to have broken only a few tests in Panel and none in HoloViews. The two unit tests that broke in Panel are https://github.com/holoviz/panel/blob/df0b277dca2fd54dd2ae016159adc5c33b61e358/panel/tests/widgets/test_base.py#L25 and https://github.com/holoviz/panel/blob/df0b277dca2fd54dd2ae016159adc5c33b61e358/panel/tests/test_viewable.py#L40, I believe they test that the `__signature__` of some Panel objects is actually *not* modified, as there is some code in Panel that modifies the signature of components. That code in Panel by the way doesn't seem to be tested, given https://github.com/holoviz/panel/blob/df0b277dca2fd54dd2ae016159adc5c33b61e358/panel/tests/conftest.py#L32. Anyway, it seems to me the change done in this PR would not affect Panel, let's double check that however if it's not tested.

I have chosen not to even try to handle custom `__init__` in 2) as I'm not even sure it's possible (@jlstevens raised some good points in the issue) and I believe that a better approach for Param in the future would be to add Parameter attributes to control how the Parameter appears in the signature (e.g. positional, hidden) and provide a post init hook so that users don't have to create an `__init__` in most cases.

### CODE REVIEW

File: param/parameterized.py [hunk 1]
```diff
@@ -3129,30 +3129,16 @@ def __set_name(mcs, name, dict_):
...
-
-        keywords = [el for grp in reversed(keyword_groups) for el in grp]
-        class_docstr = "\n"+mcs.__doc__ if mcs.__doc__ else ''
-        signature = "params(%s)" % (", ".join(keywords))
-        description = param_pager(mcs) if (docstring_describe_params and param_pager) else ''
-        mcs.__doc__ = signature + class_docstr + '\n' + description
+        if not docstring_describe_params or not param_pager:
+            return
+        class_docstr = mcs.__doc__ if mcs.__doc__ else ''
+        description = param_pager(mcs)
```
Comment (philippjfr) [inline param/parameterized.py:3140]: I actually really dislike having the pager in the docstring too, but that's a separate discussion I guess.

File: param/parameterized.py [hunk 2]
```diff
@@ -3185,6 +3171,42 @@ def __is_abstract(mcs):
...
+            for k, v in sorted(cls.__dict__.items()):
+                if isinstance(v, Parameter) and k not in processed_kws and not v.readonly:
+                    keyword_group.append(k)
+                    processed_kws.add(k)
+            keyword_groups.append(keyword_group)
+
+        keywords = [el for grp in reversed(keyword_groups) for el in grp]
+        # if mcs.name == 'C':
+        #     import pdb; pdb.set_trace()
+        # keywords = [el for grp in keyword_groups for el in grp]
```
Comment (maximlt) [inline param/parameterized.py:3201]:
```suggestion
```

**Comment by philippjfr** [comment]
Made some minor changes:

- Pull out the default signature into a global variable
- Cache the generated signature on `._param__private`

File: tests/testparameterizedobject.py [hunk 3]
```diff
@@ -1411,3 +1412,143 @@ def __init__(self, **params):
             super().__init__(**params)
 
     P()
+
+
+def test_parameterized_docstring():
+    # TODO
+    pass
+
```
Comment (philippjfr) [inline tests/testparameterizedobject.py:1420]:
```suggestion
def test_parameterized_docstring():
    # TODO
    pass

```

**Review by philippjfr — APPROVED** [review]
Looks good and I'm happy with this conservative approach to auto-generating signatures!

File: param/parameterized.py (hunk 1 above)
Comment (jbednar) [reply param/parameterized.py:3141]: Same -- I want to control the pager myself, using Emacs in my case, and am annoyed when any single program tries to handle paging.

**Pull request merged**